Automattic/jetpack
# Upgrade Flow: Connecting Jetpack users

#### Changes proposed in this Pull Request:
If a Jetpack user is not connected to WP.com, clicking on the "Upgrade" button will lead them nowhere.
This PR introduces the changes to let such users register/authenticate on WP.com (if needed), and link their account to the Jetpack site.

#### Jetpack product discussion
p9dueE-1M4-p2

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:

**Due to changes in https://github.com/Automattic/jetpack/pull/18548, this PR (as well as the `master` branch itself) outputs error notices.
You'll need to disable `WP_DEBUG` for redirects to work properly.**

1. Connect a Jetpack site with a Free plan, create an additional WP user.
2. Log in as that user in another browser or a private mode. If you're logged in WP.com in that browser, please logout.
3. Click an "Upgrade" button in Jetpack Dashboard or Jetpack Settings. You should be redirected to the Calypso connection flow.
4. Authenticate into WP.com. You should be redirected to the WP.com checkout page with that product in your cart.
5. Go back to Jetpack Dashboard, confirm that you are connected as a secondary user. Unlink that user from WP.com.
6. Click on any other "Upgrade" button. You should be redirected to the Calypso connection flow being logged into WP.com.
7. After clicking "Approve", you should get redirected to WP.com checkout with that product in the cart (previous "upgrade" product will probably still be in the cart too).
8. Go back to Jetpack Dashboard, confirm that you're connected to WP.com as a secondary user.

#### Proposed changelog entry for your changes:
* Improved product upgrade flow for non-connected Jetpack users.
